File: Part04/Week02/tsp.py
import math
import cProfile

class Solution():
    def __init__(self, filename) -> None:
        self.n, self.vertices = self.read_file(filename)
    
    @staticmethod
    def read_file(filename):
        with open(filename) as f:
            lines = f.readlines()
            n = int(lines[0])
            vertices = [(float(line.split()[0]), float(line.split()[1])) for line in lines[1:]]
            return n, vertices
    
    @staticmethod
    def next_set_of_n_elements(x):
        if x == 0:
            return x
        smallest = x & (-x)
        ripple = x + smallest
        new_smallest = ripple & (-ripple)
        ones = ((new_smallest // smallest) >> 1) - 1
        return ripple | ones
    
    def dist(self, i, j):
        x, y  = self.vertices[i], self.vertices[j]
        return math.sqrt((x[0] - y[0]) ** 2 + (x[1] - y[1]) ** 2)
    
    def tsp(self):
        # source: n - 1
        only_one = 1 << (self.n -1)

        dp = [{} for _ in range(self.n)]
        dp[self.n - 1][only_one] = 0

        for m in range(0, self.n - 1):
            new_dp = [{} for _ in range(self.n)]
            s = only_one | ((1 << (m+1)) - 1)
            while s < (1 << self.n):
                for j in range(0, self.n):
                    # check if j in S:
                    with_j = 1 << j
                    if s & with_j != 0:
                        # dp[j][s] = float('inf')
                        new_dp[j][s] = float('inf')
                        for k in range(0, self.n):
                            # check if k in S
                            with_k = 1 << k
                            if s & with_k != 0 and k != j:
                                # check if S-{j} in dp[k]:
                                if s & (~with_j) in dp[k]:
                                    # dp[j][s] = min(dp[j][s], dp[k][s & (~with_j)] + self.dist(j, k))
                                    new_dp[j][s] = min(new_dp[j][s], dp[k][s & (~with_j)] + self.dist(j, k))
                s = self.next_set_of_n_elements(s)
                dp = new_dp

        result = float('inf')
        all_set = (1 << self.n) - 1
        for j in range(0, self.n - 1):
            result = min(result, dp[j][all_set] + self.dist(self.n-1, j))
        return result
        

# solution = Solution('/mnt/e/practice/h_programming_basis/Standford_Algorithm/Part04/Week02/test.txt')
solution = Solution('/mnt/e/practice/h_programming_basis/Standford_Algorithm/Part04/Week02/tsp.txt')
cProfile.run('print(int(solution.tsp()))')

File: Part04/Week02/cpp_impl/src/tsp.cpp
#include<fstream>
#include<cstdint>
#include<cmath>
#include<cstring>

using std::string;  

float ** readfile(string filename, uint32_t* np) {
    std::ifstream input(filename);
    uint32_t n;
    input >> n;
    *np = n;

    float vx[n];
    float vy[n];
    for(uint32_t i = 0; i < n; i++) {
        input >> vx[i] >> vy[i];
    }
    float** dist = new float*[n];
    for (uint32_t i = 0; i < n; i++) {
        dist[i] = new float[n];
    }
    for (uint32_t i = 0; i < n; i++) {
        for (uint32_t j = 0; j < n; j++) {
            float square = (vx[i] - vx[j]) * (vx[i] - vx[j]) + (vy[i] - vy[j]) * (vy[i] - vy[j]);
            dist[i][j] = sqrtf32(square);
        }
    }
    return dist;
}

unsigned next_set_of_n_elements(unsigned x) 
  {
     unsigned smallest, ripple, new_smallest, ones;
   
     if (x == 0) return 0;
     smallest     = (x & -x);
     ripple       = x + smallest;
     new_smallest = (ripple & -ripple);
     ones         = ((new_smallest/smallest) >> 1) - 1;
     return ripple | ones;
}
float min(float x, float y) {
    return x < y? x: y;
}

float tsp(uint32_t n, float** dist) {
    uint32_t only_one = 1 << (n - 1);
    uint32_t maxones = 1 << n;
    float **dpp;
    
    dpp = new float*[n];
    
    for (uint32_t i = 0; i < n; i++) {
        dpp[i] = new float[maxones];
        for (uint32_t s = 0; s < maxones; s++) {
            dpp[i][s] = INFINITY;
        }
    }
    dpp[n - 1][only_one] = 0;

    for(uint32_t m = 0; m < n - 1; m++) {
        for (uint32_t s = only_one | (((1 << (m+1)) - 1)); s < maxones; s = next_set_of_n_elements(s)) {
            for(uint32_t j = 0; j < n; j++) {
                uint32_t with_j = 1 << j;
                if ((s & with_j) != 0) {
                    for(uint32_t k = 0; k < n; k++) {
                        uint32_t with_k = 1 << k;
                        if (((s & with_k) != 0) && k != j) {
                            dpp[j][s] = min(dpp[j][s], dpp[k][s & (~with_j)] + dist[j][k]);
                        }
                    }
                }
            }
        }
    }

    float result = INFINITY;
    // printf("%f\n", result);
    for (uint32_t j = 0; j < n - 1; j++) {
        result = min(result, dpp[j][(1 << n) - 1] + dist[j][n-1]);
    }
    for (uint32_t i = 0; i < n; i++) {
        delete[] dpp[i];
    }
    delete[] dpp;
    return result;
}
int main() {
    uint32_t n;
    float** dist;
    dist = readfile("/mnt/e/practice/h_programming_basis/Standford_Algorithm/Part04/Week02/tsp.txt", &n);
    printf("%f\n", tsp(n, dist));
    for (uint32_t i = 0; i < n; i++) {
        delete[] dist[i];
    }
    delete[] dist;
    return 0;
}